rfcs/2022-07-28-13691-registered-internal-events.md

```diff
@@ -0,0 +1,231 @@
+    type Data;
+    fn emit(&self, data: Data);
+}
+```
+
+To assist with providing the necessary types for emitting the registered events, a set of common
+unit structures will be defined for the required data:
+
+```rust
+struct ByteSize(usize);
+
+struct ByteSizeCount(usize, usize);
+```
+
+Finally, here is a sample implementation of the above traits, taken from the existing
+`EndpointBytesReceived` internal event:
+
+```rust
+use metrics::Counter;
+
+struct RegisterEndpointBytesReceived {
+    bytes_total: Counter,
+    protocol: &'static str,
+    endpoint: String,
+}
+
+struct EndpointBytesReceivedHandle {
+    bytes_total: Counter,
+    protocol: &'static str,
+    endpoint: String,
+}
+
+impl RegisterInternalEvent for RegisterEndpointBytesReceived {
+    type Handle = EndpointBytesReceivedHandle;
+
+    fn register(self) -> Self::Handle {
+        let bytes_total = register_counter!(
+            "component_received_bytes_total",
+            "protocol" => self.protocol,
+            "endpoint" => self.endpoint.clone(),
+        );
+        Self {
+            bytes_total,
+            protocol: self.protocol,
+            endpoint: self.endpoint,
+        }
+    }
+}
+
+impl InternalEventHandle for EndpointBytesReceivedHandle {
+    type Data = ByteSize;
+    fn emit(&self, data: ByteSize) {
+        trace!(
+            message = "Bytes received.",
+            byte_size = %data.0,
+            protocol = %self.protocol,
+            endpoint = %self.endpoint,
+        );
+        self.bytes_total.add(data.0);
+    }
+}
+
+// In component code:
+
+use crate::internal_events::{InternalEventHandle, RegisterEndpointBytesReceived};
+
+let handle = register!(RegisterEndpointBytesReceived {
+    protocol = "https",
+    endpoint = self.config.endpoint.clone(),
+);
+
+handle.emit(ByteSize(received.len()));
+```
+
+## Rationale
+
+- This interface for registered internal event handles is notationally equivalent to the existing
+  `InternalEvent` interface, using field names and named types for all data.
+
+## Drawbacks
+
+- This increases the complexity of the internal event interface, presenting two different interfaces
+  for internal events.
+
+## Alternatives
+
+The simplest method of providing this is to just write new internal event structures that are
+created through a struct method returning `Self`. This would allow us to name the handle more
+simply, but have the downside of not allowing for naming the creation parameters and not enforcing
+the register pattern at check/compile time on all such events.
+
+```rust
+pub struct BytesSentHandle { … }
+
+impl BytesSentHandle {
+    fn new(protocol: &str, etc: &str) -> Self { … }
+}
+```
+
+Alternately, we could audit our metric usage and ensure that all hot metrics are emitted in batches
+instead of per-event or equivalent. This is not easily determined just by examining the code,
+requiring run-time analysis. Also, batching emission may not be straightforward neither.
+
```

Comment:
I wonder: if we're going to have to implement the a new type/trait for registerable metrics, would it be worth it to write inherent methods on the handle type that take each necessary parameter vs a single value?

That way instead of it being a generic `emit(data_val)`, it would be more like `track_bytes(byte_len)`, or some such method, naming/arguments tailored to the logical event. This would, however, invalidate my above idea about doing implicit trait impls for common usages like byte size/count from a `&EventArray`, etc.

Reply:
I think this only works when discussing the standard metrics (`count` and `byte_size`, say). I think we want this design to work beyond those simple cases to include any kind of per-event measurement such as end-to-end latency or query response times, etc.

Reply:
Yeah, I mean, I don't think doing so precludes that from happening... it would just be that the emission method(s) would be written by hand, but that there could be more than _one_.

Was just a random thought, though, not necessarily a strong vision of what it should look like.